README.md
# Flask-Admin Dashboard Network Automation App

Basic dashboard app with Admin LTE template and Flask Admin, it has:

- User Registration
- Login as general or admin user
- Roles management
- Create form in modal window by default
- Inline editing enabled by default
- Skins and  layout customization
- Templates for a number of protocols for CISCO devices
- Python automation scripts using Paramiko, Netmiko, NAPALM and Ansible
 
Utilities: 

  - AdminLTE Bootstrap template
  - Flask-Security
  - Flask-Admin
  - SQLite
  - Customized templates for protocols parameters

### Screenshots
![Diagram.png](./image.png)
![FormExample.png](./image-1.png)
![FormOutput.png](./image-2.png)
![Files.png](./image-3.png)



**I hope you enjoy it.**


history.txt
    1  jourtnalctl -f | grep bash
    2  journalctl -f | grep bash
    3  sudo lsof -i -P -n
    4  sudo lsof -i -P -n | grep LISTEN 
    5  pkill -9 8942
    6  sudo lsof -i -P -n | grep LISTEN 
    7  pkill -9 8973
    8  sudo lsof -i -P -n | grep LISTEN 
    9  cd /var/www/html/
   10  ls
   11  cd licenta/
   12  ls
   13  sudo nano app.
   14  sudo nano app.py 
   15  python3 app.py 
   16  sudo nano app.py 
   17  python3 app.py 
   18  sudo nano app.py 
   19  python3 app.py 
   20  sudo nano app.py 
   21  python3 app.py 
   22  sudo nano app.py 
   23  python3 app.py 
   24  gunicorn --bind 0.0.0.0:8888 wsgi:app
   25  source venv/bin/activate
   26  gunicorn --bind 0.0.0.0:8888 wsgi:app
   27  deactivate 
   28  sudo nano /etc/systemd/system/flask.service 
   29  rm app.sock 
   30  ls
   31  sudo rm app.sock 
   32  ls
   33  sudo systemctl start flask
   34  sudo systemctl daemon-reload
   35  sudo systemctl start flask
   36  sudo systemctl enable flask
   37  sudo systemctl status flask
   38  sudo nano /etc/systemd/system/flask.service 
   39  sudo systemctl daemon-reload
   40  sudo systemctl start flask
   41  sudo systemctl enable flask
   42  sudo systemctl status flask
   43  sudo nano /etc/nginx/conf.d/flask.conf 
   44  sudo systemctl restart nginx
   45  sudo systemctl daemon-reload
   46  sudo systemctl start flask
   47  sudo systemctl status flask
   48  cd /home
   49  cd osboxes/
   50  ls
   51  cd licenta/
   52  ls
   53  source venv/bin/activate
   54  pip install uwsgi
   55  python3 app.py 
   56  lsof -i 5000
   57  lsof -i:5000
   58  sudo nano app.py 
   59  python3 app.py 
   60  sudo nano app.py 
   61  python3 app.py 
   62  sudo nano wsgi.py 
   63  sudo nano app.ini
   64  ls
   65  sudo nano app.ini
   66  uwsgi --socket 0.0.0.0:5005 --protocol=http -w wsgi:app
   67  deactivate
   68  sudo nano app.ini
   69  ls
   70  sudo rm app.sock 
   71  ls
   72  sudo nano /etc/systemd/system/flask.service 
   73  sudo systemctl start flask
   74  sudo systemctl daemon-reload
   75  sudo systemctl start flask
   76  sudo systemctl enable flask
   77  sudo systemctl status flask
   78  sudo nano /etc/systemd/system/flask.service 
   79  sudo nano app.ini
   80  nano app.sock
   81  ls
   82  sudo nano app.ini
   83  sudo systemctl restart nginx
   84  sudo systemctl restart flask
   85  sudo systemctl status flask
   86  sudo rm -r /var/www/html/licenta/
   87  unlink /etc/nginx/sites-available/flask 
   88  sudo unlink /etc/nginx/sites-available/flask 
   89  sudo unlink /etc/nginx/sites-enabled/flask 
   90  ls /etc/nginx/sites-available/
   91  ls /etc/nginx/sites-enabled/
   92  sudo systemctl restart nginx
   93  sudo systemctl restart flask
   94  sudo systemctl status flask
   95  sudo systemctl status nginx
   96  sudo nano /etc/systemd/system/flask.service 
   97  sudo systemctl status flask
   98  sudo systemctl restart flask
   99  sudo systemctl daemon-reload
  100  sudo systemctl restart flask
  101  sudo systemctl status flask
  102  sudo nano /etc/nginx/sites-available/flask
  103  sudo rm /etc/nginx/conf.d/flask.conf 
  104  sudo nano /etc/nginx/sites-available/flask
  105  sudo rm /etc/nginx/sites-available/flask
  106  sudo nano /etc/nginx/sites-available/flask
  107  ls
  108  server {
  109  sudo ln -s /etc/nginx/sites-available/flask /etc/nginx/sites-enabled
  110  ln -s /etc/nginx/sites-available/flask /etc/nginx/sites-enabled
  111  sudo nginx -t
  112  sudo systemctl restart nginx
  113  sudo ufw delete allow 5005
  114  sudo ufw delete allow 5000
  115  sudo ufw allow 'Nginx Full'
  116  sudo nano /etc/nginx/sites-available/flask
  117  ls
  118  nano app.sock 
  119  sudo nano /etc/host
  120  sudo nano /etc/hosts
  121  sudo systemctl restart nginx
  122  sudo systemctl restart flask
  123  sudo systemctl enable flask
  124  sudo systemctl status flask
  125  sudo less /var/log/nginx/error.log
  126  sudo nano /etc/nginx/nginx.conf 
  127  sudo systemctl restart nginx
  128  sudo systemctl restart flask
  129  sudo systemctl enable flask
  130  sudo systemctl status flask
  131  sudo less /var/log/nginx/error.log
  132  sudo less /var/log/nginx/access.log
  133  sudo journalctl -u nginx
  134  sudo journalctl -u flask
  135  sudo systemctl status flask
  136  sudo nano /etc/nginx/sites-available/flask
  137  sudo systemctl restart nginx
  138  sudo systemctl restart flask
  139  sudo systemctl daemon-reload
  140  sudo systemctl restart flask
  141  sudo systemctl enable flask
  142  sudo systemctl status flask
  143  sudo nano /etc/nginx/sites-available/flask
  144  sudo unlink /etc/nginx/sites-enabled/flask 
  145  sudo ln -s /etc/nginx/sites-available/flask /etc/nginx/sites-enabled 
  146  sudo nginx -t
  147  sudo systemctl restart nginx
  148  sudo nano /etc/nginx/sites-available/flask
  149  sudo nano /etc/nginx/sites-enabled/flask 
  150  sudo nano /etc/nginx/sites-available/flask
  151  sudo nano /etc/nginx/sites-enabled/flask 
  152  y
  153  sudo nano /etc/nginx/sites-enabled/flask 
  154  sudo nano /etc/nginx/sites-available/flask
  155  sudo systemctl restart nginx
  156  sudo nano /etc/hosts
  157  sudo systemctl restart nginx
  158  ls
  159  nano templates/myscripts/ping.py 
  160  cd ..
  161  ls
  162  pin 192.168.122.77
  163  ping 192.168.122.77
  164  nano testping.py
  165  python3 testping.py 
  166  nano testping.py
  167  cd licenta/
  168  python3 app.py
  169  uwsgi --socket 0.0.0.0:5005 --protocol=http -w wsgi:app
  170  sudo nano app.ini 
  171  sudo rm app.ini 
  172  sudo nano flask.ini 
  173  sudo nano /etc/systemd/system/flask.service 
  174  sudo nano wsgi.py 
  175  sudo nano /etc/systemd/system/flask.service 
  176  sudo systemctl daemon-reload
  177  sudo systemctl start flask
  178  sudo systemctl enable flask
  179  sudo systemctl status flask
  180  sudo systemctl restart flask
  181  sudo systemctl enable flask
  182  sudo systemctl status flask

templates/myscripts/autosecure.py
import threading
import os
import tempfile
import sys
import io
from subprocess import run
import requests
from netmiko import ConnectHandler, NetMikoTimeoutException, NetMikoAuthenticationException
from datetime import datetime
import uuid

# from templates.myscripts import netmiko_check_autosecure_config, napalm_check_connectivity
from templates.myscripts import netmiko_check_autosecure_config

old_stdout = sys.stdout
new_stdout = io.StringIO()
sys.stdout = new_stdout
output = ""
ROOT_DIR = os.path.dirname(os.path.abspath(__file__))

# Interactive full session of AutoSecure
def autosecure_full(type, device, tmp_file, vars_array):
    fout = open(tmp_file, "wt")

    fout.write("auto secure " + type + '\n')

    # Is this router connected to internet? [no]
    fout.write("no" + '\n')

    # Enter the security banner {Put the banner between
    # k and k, where k is any character}
    fout.write("k" + vars_array[0] + "k" + '\n')

    # Enter the new enable secret
    # Choose a secret that's different from password
    # Invalid Password length - must contain 6 to 25 characters. Password configuration failed
    fout.write(vars_array[1] + '\n')

    # Confirm the enable secret
    fout.write(vars_array[1] + '\n')

    # Enter the new enable password
    # Choose a password that's different from secret
    fout.write(vars_array[2] + '\n')

    # Confirm the enable password
    fout.write(vars_array[2] + '\n')


    if netmiko_check_autosecure_config.autosecure_local_user(device) == 0:
        # Configuration of local user database
        # Enter the username
        fout.write(vars_array[3] + '\n')

        # Enter the password:
        # % Invalid Password length - must contain 6 to 25 characters. Password configuration failed
        fout.write(vars_array[4] + '\n')

        # Confirm the password
        fout.write(vars_array[4] + '\n')

    # Blocking Period when Login Attack detected
    fout.write(vars_array[5] + '\n')

    # Maximum Login failures with the device
    fout.write(vars_array[6] + '\n')

    # Maximum time period for crossing the failed login attempts
    fout.write(vars_array[7] + '\n')

    # Configure SSH server?
    fout.write(vars_array[8] + '\n')

    if vars_array[8] == 'yes':
        # Enter the hostname
        fout.write(vars_array[9] + '\n')
        # Enter the domain-name
        fout.write(vars_array[10] + '\n')

    if netmiko_check_autosecure_config.autosecure_firewall(device) == 0:
        # Configure CBAC Firewall feature? [yes/no]
        fout.write(vars_array[11] + '\n')

        if vars_array[11] == "no":
            # Enable tcp intercept feature? [yes/no]
            fout.write(vars_array[12] + '\n')

    # Apply this configuration to running-config? [yes]
    fout.write("yes" + '\n')

    fout.close()


# Non-interactive session of AutoSecure
# empty vars array
def autosecure_nointeract(type, tmp_file):
    fout = open(tmp_file, "wt")

    fout.write("auto secure " + type + '\n')

    fout.close()


# AutoSecure TCP Intercept
# empty vars array
def autosecure_tcpintercept(type, tmp_file):
    fout = open(tmp_file, "wt")

    fout.write("auto secure " + type + '\n')

    # Is this router connected to internet? [no]
    fout.write("no" + '\n')

    # Enable tcp intercept feature? [yes/no]
    fout.write("yes" + '\n')

    # Apply this configuration to running-config? [yes]
    fout.write("yes" + '\n')

    fout.close()


# AutoSecure NTP
# empty vars array
def autosecure_ntp(type, tmp_file):
    fout = open(tmp_file, "wt")

    fout.write("auto secure " + type + '\n')

    # Is this router connected to internet? [no]
    fout.write("no" + '\n')

    # Apply this configuration to running-config? [yes]
    fout.write("yes" + '\n')

    fout.close()


# AutoSecure SSH
# la nivel de interfata grafica tb o verificare asemanatoare pt a nu permite utilizatorului sa seteze domain name
def autosecure_ssh(type, device, tmp_file, vars_array):
    fout = open(tmp_file, "wt")

    fout.write("auto secure " + type + '\n')

    # Is this router connected to internet? [no]
    fout.write("no" + '\n')

    # Configure SSH server? [yes]
    fout.write("yes" + '\n')

    # Enter the domain-name
    if netmiko_check_autosecure_config.autosecure_domain_name(device) == 0:
        fout.write(vars_array[10] + '\n')

    # Apply this configuration to running-config? [yes]
    fout.write("yes" + '\n')

    fout.close()


# AutoSecure Login
def autosecure_login(type, device, tmp_file, vars_array):
    fout = open(tmp_file, "wt")

    fout.write("auto secure " + type + '\n')

    # Is this router connected to internet? [no]
    fout.write("no" + '\n')

    if netmiko_check_autosecure_config.autosecure_login(device) == 0:
        # Enter the new enable secret
        # Choose a secret that's different from password
        # Invalid Password length - must contain 6 to 25 characters. Password configuration failed
        fout.write(vars_array[1] + '\n')

        # Confirm the enable secret
        fout.write(vars_array[1] + '\n')

        # Enter the new enable password
        # Choose a password that's different from secret
        fout.write(vars_array[2] + '\n')

        # Confirm the enable password
        fout.write(vars_array[2] + '\n')


    if netmiko_check_autosecure_config.autosecure_local_user(device) == 0:
        # Configuration of local user database
        # Enter the username
        fout.write(vars_array[3] + '\n')

        # Enter the password:
        # % Invalid Password length - must contain 6 to 25 characters. Password configuration failed
        fout.write(vars_array[4] + '\n')

        # Confirm the password
        fout.write(vars_array[4] + '\n')

    # Blocking Period when Login Attack detected
    fout.write(vars_array[5] + '\n')

    # Maximum Login failures with the device
    fout.write(vars_array[6] + '\n')

    # Maximum time period for crossing the failed login attempts
    fout.write(vars_array[7] + '\n')

    # Apply this configuration to running-config? [yes]
    fout.write("yes" + '\n')

    fout.close()


# AutoSecure Firewall
def autosecure_firewall(type, tmp_file):
    fout = open(tmp_file, "wt")

    fout.write("auto secure " + type + '\n')

    # Is this router connected to internet? [no]
    fout.write("no" + '\n')

    # Configure CBAC Firewall feature? [yes/no]
    fout.write('yes' + '\n')

    # Apply this configuration to running-config? [yes]
    fout.write("yes" + '\n')

    fout.close()


def check_autosecure_type(device, connection, tmp_file, vars_array):
    copy_array = vars_array[1:]
    if vars_array[0] == "management full":
        type = "management full"
        autosecure_full(type, device, tmp_file, copy_array)
        return 1
    elif vars_array[0] == "management no-interact":
        type = "management no-interact"
        autosecure_nointeract(type, tmp_file)
        return 1
    elif vars_array[0] == "management ssh":
        type = "management ssh"
        autosecure_ssh(type, device, tmp_file, copy_array)
        return 1
    elif vars_array[0] == "management ntp":
        type = "management ntp"
        autosecure_ntp(type, tmp_file)
        return 1
    elif vars_array[0] == "management login":
        type = "management login"
        autosecure_login(type, device, tmp_file, copy_array)
        return 1
    elif vars_array[0] == "forwarding full":
        type = "forwarding full"
        autosecure_firewall(type, tmp_file)
        return 1
    elif vars_array[0] == "forwarding no-interact":
        type = "forwarding no-interact"
        autosecure_nointeract(type, tmp_file)
        return 1
    elif vars_array[0] == "forwarding tcp-intercept":
        type = "forwarding tcp-intercept"
        autosecure_tcpintercept(type, tmp_file)
        return 1
    elif vars_array[0] == "forwarding firewall":
        type = "forwarding firewall"
        autosecure_firewall(type, tmp_file)
        return 1
    elif vars_array[0] == "firewall":
        type = "firewall"
        autosecure_firewall(type, tmp_file)
        return 1
    elif vars_array[0] == "full":
        type = "full"
        autosecure_full(type, device, tmp_file, copy_array)
        return 1
    elif vars_array[0] == "login":
        type = "login"
        autosecure_login(type, device, tmp_file, copy_array)
        return 1
    elif vars_array[0] == "no-interact":
        type = "no-interact"
        autosecure_nointeract(type, tmp_file)
        return 1
    elif vars_array[0] == "ntp":
        type = "ntp"
        autosecure_ntp(type, tmp_file)
        return 1
    elif vars_array[0] == "ssh":
        type = "ssh"
        autosecure_ssh(type, device, tmp_file, copy_array)
        return 1
    elif vars_array[0] == "tcp_intercept":
        type = "tcp_intercept"
        autosecure_tcpintercept(type, tmp_file)
        return 1
    else:
        return 0


def send_commands(connection, tmp_file):
    f = open(tmp_file, "r")
    for line in f:
        result = connection.send_command_timing(line, 1, 150, True, True, True, False, None, False, None, False, False)
        print("START RESULT")
        print(result)
        print("END RESULT")
        output = result

    f.close()


def autosecure(device, vars_array):
    connection = ConnectHandler(**device)

    print('Entering enable mode...')
    connection.enable()

    # create a temporary file to create the config
    tmp_file = str(uuid.uuid4())
    try:
        if check_autosecure_type(device, connection, tmp_file, vars_array) != 0:
            print('Sending commands from file...')
            send_commands(connection, tmp_file)

    finally:
        print('Done!')
        path = tmp_file
        CONFIG_PATH = os.path.join(ROOT_DIR, path)
        os.remove(path)

    print('Closing connection...')
    connection.disconnect()

def do_autosecure(devices, os_type, username, password, enable, vars_array):
    # mylist = devices.split(', ')
    return_statement = ""

    for ip in devices:
        if os_type == 'ios':
            os_type = 'cisco_ios'

        cisco_device = {
                'device_type': os_type,
                'host': ip,
                'username': username,
                'password': password,
                'port': 22,  # optional, default 22
                'secret': enable,  # this is the enable password
                'verbose': True  # optional, default False
        }
        # note that the enable secret and passwords are only sent once, the confirmation is interface-based
        # vars_array = ['1', 'knot setk', 'not set', 'not set 2', 'not set', 'not set 3', '30', '30', '30', 'No', 'not set', 'not set', 'No', 'No']
        try:
            autosecure(cisco_device, vars_array)
            return_statement += "Succes for " + ip + ".\n"
        except ConnectionRefusedError as err:
            this_error = f"Connection Refused: {err}\n"
            return_statement += this_error
        except TimeoutError as err:
            this_error = f"Connection Refused: {err}\n"
            return_statement += this_error
        except Exception as err:
            this_error = f"Oops: {err}\n"
            return_statement += this_error

    return return_statement


# vars_array = ['firewall', 'knot setk', 'not set', 'not set 2', 'not set', 'not set 3', '30', '30', '30', 'No', 'not set',
#               'not set', 'No', 'No']
#
# do_autosecure('192.168.204.16', 'ios', 'admin', 'cisco', 'parola', vars_array)